## TESTS FOR RATITE-SPECIFIC ACCELERATION, ETC ##

#for each chicken-ratite pair, get the alignability including duplicates for that pair (should be 0, 1, 2+)
#run in alignDepths subdir
for $SP in droNov casCas aptHaa aptOwe aptRow rheAme rhePen strCam
do
	halAlignmentDepth --inMemory --countDupes --noAncestors --targetGenomes $SP ~/ratite_scratch/wga/ratite_final_20150627/ratiteAlign.hal galGal > galGal-$SP.align.wig &
done

#everything below is run in the accelTests subdir

#setup:
#make galGal chrom list for downstream processing
cut -f1,1 ../galGal.chromsizes | sort -u > chr.list

#get beds
cp ../ce_beds/final_ces_noratite.tree2.bed.gz final_noratite_tree2.bed.gz
cp ../ce_beds/final_ces.tree2.bed.gz final_all_tree2.bed.gz
gunzip *.bed.gz

#1:
mkdir all_species
cd all_species
mkdir -p ratite tips clades tinamou
NEUTMOD="$HOME/ratite_scratch/wga/phast/neutMods/neut_ver2_final.named.mod"
BED="../final_all_tree2.bed"
for CHR in $(cat ../chr.list)
do
	ALIGN="$HOME/ratite_scratch/wga/phast/final_mafs/$CHR.ss"
	if [[ -f $ALIGN ]]
	then
		grep "^$CHR" $BED > $CHR.temp.bed
		sbatch ../est_accel.sh $NEUTMOD $CHR $ALIGN
	fi
done
cd ..

#aslo run: rr_only subset, no ratite alignments basal Paleo only subset 
#need to add code for these from subdirs

##NOTE for CHR = NW_003764425.1 the following command was used since notPer is missing from alignment
phyloP --method LRT --features $CHR.temp.bed --mode ACC --branch cryCin,tinGut,cryCin-tinGut,eudEle,cryCin-eudEle $NEUTMOD $ALIGN > tinamou/$CHR.Tinamou.out
#and for basal palaeognath, no ratite test
phyloP --method LRT --features $CHR.temp.bed --mode ACC --branch aptHaa-strCam $NEUTMOD $ALIGN > tinamou/$CHR.basalPaleo.out


#clean up
#make headers
head -n 1 ratite/NC_006088.3.Ratite.out > all_ratite.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_Casuar.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_Rhea.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_Kiwi.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_tinamou.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_casCas.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_droNov.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_strCam.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_rheAme.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_rhePen.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_aptHaa.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_aptOwe.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_aptRow.out
head -n 1 ratite/NC_006088.3.Ratite.out > all_basalPaleo.out

for FILE in $(ls ratite/*.Ratite.out)
do
	tail -n +2 $FILE >> all_ratite.out
done

for FILE in $(ls tinamou/*.Tinamou.out)
do
	tail -n +2 $FILE >> all_tinamou.out
done

for FILE in $(ls tinamou/*.basalPaleo.out)
do
	tail -n +2 $FILE >> all_basalPaleo.out
done

for CLADE in Rhea Casuar Kiwi
do
	for FILE in $(ls clades/*.$CLADE.out)
	do
		tail -n +2 $FILE >> all_${CLADE}.out
	done
done

for TIP in casCas droNov strCam rheAme rhePen aptHaa aptOwe aptRow
do
	for FILE in $(ls tips/*.$TIP.out)
	do
		tail -n +2 $FILE >> all_${TIP}.out
	done
done


##BELOW HERE NOT RUN##

#general procedure for acceleration tests:
#goal is to do three things:
#0) should implement some missing-data filters and some duplication filters
#1) run standard phyloP to get test of acceleration for each ratite branch, all ratite branches together, and the tinamou clade
#2) run standard phyloP with the same set of branches but on a set of 5e6 random alignments generated by msa_view
#--the tricky part here is matching the length distribution
#3) run phyloP with SPH option to estimate posterior distribution of sites?

#0: filter beds
#use alignability to make sure element is present in half of species
#use coverage to make sure element is not duplicated in any ratites
#use coverage to make sure element is present in at least 50% of ratites

#0b: set up random sampled branches
##generate a null model with 10 random samples of lineages
#get named tree file

for i in 1 2 3 4 5 6 7 8 9 10
do
	brinput=$(nw_labels named_tree.nh | grep -v "anoCar" | chooseLines -k 13 - | tr '\n' ',')
	echo $brinput > rand$i.branches	
done

#clean up random sets




#2:
#first need to make fake alignments
#this code gets a SS alignment of all the conserved elements
mkdir -p random
cd random
BED="../final_noratite_tree2.bed"
for CHR in $(cat ../chr.list)
do
	ALIGN="../../final_mafs/$CHR.ss"
	if [[ -f $ALIGN ]]
	then
		grep "^$CHR" $BED > $CHR.temp.bed
		msa_split $ALIGN --by-category --features $CHR.temp.bed --in-format SS --out-format SS --out-root $CHR.cons
	fi
done
SPLIST="galGal,pseHum,taeGut,ficAlb,corBra,melUnd,falPer,picPub,lepDis,halLeu,aptFor,pygAde,fulGla,nipNip,balReg,chaVoc,calAnn,chaPel,cucCan,colLiv,mesUni,melGal,anaPla,aptHaa,aptOwe,aptRow,casCas,droNov,cryCin,tinGut,eudEle,notPer,rheAme,rhePen,strCam,allMis,allSin,croPor,gavGan,chrPic,cheMyd,anoCar"
FILELIST=$(ls *.cons.bed_feature-1.ss)
msa_view --aggregate $SPLIST $FILELIST --out-format SS > all.cons.ss
#now read this in R and generate the necessary bed file and sampled alignment to run phyloP
#load R
export LD_LIBRARY_PATH=""
module purge
module load R
module load pcre
export R_LIBS_USER=$HOME/sw/R:$R_LIBS_USER
R

#R script here:
library(rphast)
align<-read.msa("all.cons.ss")
elements<-read.feat("../final_all_tree2.bed")
#now we will get the lengths of the features
ele.len<-elements$end - elements$start
#now simluate 100,000 bed elements for each feature length, rounded to the nearest 5 bp
mround <- function(x,base){ 
        base*round(x/base) 
} 
ele.len.5bp<-unique(mround(ele.len,5))
nrep<-1e6
for (splitLength in ele.len.5bp) {
	simMsa<-sample.msa(align, nrep*splitLength, replace=TRUE)
	output.name.msa<-paste0(c("sampled.", splitLength, "bp.ss"), collapse="")
	output.name.gff<-paste0(c("sampled.", splitLength, "bp.gff"), collapse="")
	startIdx<-seq(from=1, by=splitLength, length.out=nrep)
	simFeat<-feat(seqname=names.msa(simMsa)[1], src="sim", feat=".", start=startIdx, end=startIdx+splitLength-1)
	write.feat(simFeat, file=output.name.gff)	
	write.msa(simMsa, file=output.name.msa)
}

#now can run phyloP like:
#phyloP --method LRT --features sampled.30bp.gff --mode ACC --branch rheAme $HOME/ratite_scratch/wga/phast/neutMods/neut_ver2_final.named.mod sampled.30bp.ss

